src/server/project.ts

```diff
@@ -404,7 +404,7 @@ namespace ts.server {
                 }
             });
 
-            return addRange(result, this.typingFiles) || ts.emptyArray;
+            return addRange(result, this.compilerOptions.types ? undefined : this.typingFiles) || ts.emptyArray;

```

Comment:
Did you try not adding typings file all the time as root and instead relying on module resolution and your modified type reference resolution ?

Reply:
I don’t think they would be resolved and included automatically because nothing is referencing them. I think we’d have to convert this list into a list of automatic type reference directives, essentially. I think that approach makes sense logically, but it will be more code and produce the same outcome.

Reply:
i meant did you try running some real world scenarios and asserting if typingFiles are not in the program after program is construction is complete? May be node typings will not resolve but if thats the case why would we anyways do type Acquistion when types  is set

Reply:
> I don’t think they would be resolved and included automatically because nothing is referencing them

ATA installs and gives you list of files based on unresolved import right.. so there are places where it is imported and shoul;d work no?

Reply:
Unresolved imports work, yes, but ATA also gets typings to include from package.json dependencies: https://github.com/microsoft/TypeScript/blob/46e53016e45133dd4901c4d3de62c1e249e131ae/src/jsTyping/jsTyping.ts#L224-L225

Reply:
Here’s an example. This project has no imports, yet I get global completions for React just because it’s in my package.json.

<img width="830" alt="image" src="https://user-images.githubusercontent.com/3277153/170747452-2f7e1d88-8f9a-4e31-a490-b59fad830e61.png">

Reply:
Hmm is it then that when types set to not include anything but unresolved imports in the ATA (that is in getTypingsName)

Reply:
Can you rephrase the question? 🤔 The example I’m showing has no jsconfig at all, by the way.

Reply:
Not to do packageJson lookups when types array is set ?
Will that achieve same result as this. I think it would be better to not do ATA for types if its not going to be used
Essentially discover types should ATA for types from unresolved imports only ?
https://github.com/microsoft/TypeScript/blob/ba4525202e8e647cd71b87c7711ad0c4de8348ee/src/jsTyping/jsTyping.ts#L151

Reply:
Hm, yeah, that might work.

Reply:
This makes my test with `"types": ["jquery"]` fail, but I think I might be ok with that. Maybe automatic type reference directives (from an explicit `types` list) _shouldn’t_ resolve from the global typings cache—since they’re triggered by a TS compiler option, they should resolve the same way whether you’re in the editor or using tsc.

In other words, unresolved imports can trigger special resolution behavior because imports first and foremost are written for use at runtime; bringing in the typings on top of them is an added bonus to help your experience at design time. Type reference directives (including `"types"`) are different because they have no runtime purpose; if you wrote them, you were intentionally using a TypeScript-only feature and you should do so completely by installing the types/dependencies you say you want to reference.